gc/base/WorkStack.cpp

```diff
@@ -297,3 +297,41 @@ MM_WorkStack::flushOutputPacket(MM_EnvironmentBase *env)
 	}
 }
 
+void *
+MM_WorkStack::popNoWaitFromCurrentInputPacket(MM_EnvironmentBase *env)
+{
+	void *result = NULL;
+
+	if(NULL != _inputPacket) {
+		result = _inputPacket->pop(env);
+		if (NULL == result) {
+			/* The current input packet has been used up - return it to the output list for reuse */
+			_workPackets->putPacket(env, _inputPacket);
+			_inputPacket = NULL;
+		}
+	}
+	return result;
+}
+
+bool
+MM_WorkStack::inputPacketAvailableFromWorkPackets(MM_EnvironmentBase *env)
+{
+	return _workPackets->inputPacketAvailable(env);
+}
+
+bool
+MM_WorkStack::retrieveInputPacket(MM_EnvironmentBase *env)
+{
+	_inputPacket = _workPackets->getInputPacketNoWait(env);
+	if (NULL == _inputPacket) {
+		/* If the output packet contains at least a free entry - invert the input/output */
+		if((NULL != _outputPacket) && !_outputPacket->isEmpty()) {
+			/* swap the input packet with the output packet */
+			_inputPacket = _outputPacket;
+			_outputPacket = NULL;
+			env->_workPacketStats.workPacketsExchanged += 1;
+		}
+	}
+	return (NULL != _inputPacket);
```

Comment:
this pattern is repeated a couple of more times  in this file. perhaps you could reuse this method?

Reply:
done

Reply:
expose and use WorkPackets' monitor and waitCount for CopyForwardScheme scan queue synchronization